### ISSUE TITLE
Adding a Phrase Tag layer over a Dependency layer

### ISSUE THREAD

**Issue by Kentoseth** [inception-project/inception#2685]
Hello,

Firstly thank you for making this project and fixing my other bug reports.

This isn't a bug, it is a how-to. I checked the gitter channel and it is inactive, so I am making the query here. You can close it if I should make the query somewhere else.

In this example: https://corpus.quran.com/treebank.jsp?chapter=1&verse=1

The PoS of بِسۡمِ is:

بِ = Preposition
 Noun = سۡمِ

The dependency layer is: مجرور (meaning the noun that is attached to the preposition)

They then add what looks like another PoS layer called: PP (Preposition Phrase)

And after that comes the final Dependency layer: متعلق

I have figured out how to add Layers and Tagsets. I have setup the PoS layer and the Dependency Layer(with their relevant tagsets, which I will attach below).

I cannot figure out how to add this PP beyond the Dependency Layer.

Can you suggest how?

[bismillah.txt](https://github.com/inception-project/inception/files/7486758/bismillah.txt)

[ar-layers-tags.tar.gz](https://github.com/inception-project/inception/files/7486752/ar-layers-tags.tar.gz)

**Comment by reckart** [comment]
The default dependency layer is defined on POS annotation. So, in order to create a dependency, you first create a POS annotation on two words, then you click on the annotation bubble of the first word, hold the mouse button pressed and move it over to the bubble on the second word.

Does that help?

If not, could you maybe add a screenshot that shows the current state and then explain based on that?

**Comment by reckart** [comment]
We do still watch the gitter channel and react if anybody asks as question there.

**Comment by Kentoseth** [comment]
Below is the current layout I have (there again seems to be a bug on the display of Dependencies/PoS which exists on both FireFox and Brave).

I was able to reproduce everything up to the PP bit.

![pic](https://user-images.githubusercontent.com/2989402/140608733-6680493d-1707-4c4d-8b45-ea0e74a03181.png)

Now when I try to create the "Phrase PoS" layer, I cannot stack it on top of the Dependency Layer.

My theory regarding how the Quran Corpus is doing it for that specific verse is:

1. PoS Layer
2. Dependency Layer(attached to PoS)
3. Phrase PoS Layer(attached to 2. dependency layer)
4. Dependency Layer(attached to layer 3.)

I have just attempted an alternative where 3. is not stacked above 2. and it outputs like so:

![pic2](https://user-images.githubusercontent.com/2989402/140608909-571cae9f-998d-4504-96cd-c26486e22c60.png)

Do you have any suggestions?

**Comment by reckart** [comment]
Are you using the built-in POS and dependency layers? These have restrictions with respect to stacking - but they can be used e.g. with the CoNLL-U format. Or are you open to using only custom layers? With custom layers you have more flexibility, but they are only compatible with the XMI or WebAnno-TSV formats.

Did you set up a font zoom in the annotation page preferences?

**Comment by Kentoseth** [comment]
> Are you using the built-in POS and dependency layers?

No. I had to make my own PoS Layer because it is at single letter level. It was discussed on issue: #1960 

> Or are you open to using only custom layers?

Yes.

> but they are only compatible with the XMI or WebAnno-TSV formats.

Does this have any impact when setting up layers and tagsets via Inception?

> Did you set up a font zoom in the annotation page preferences?

I didn't know about this. I will test it out.

Thanks.

**Comment by reckart** [comment]
> Did you set up a font zoom in the annotation page preferences?

I just asked this because the arabic script looks way larger than the annotations above it and the arcs are drawn strangely - so I asked about if you increased the font size in the preferences dialog because that could be causing this. Otherwise, I am not sure which what you mean by "there again seems to be a bug on the display of Dependencies/PoS".

> Does this have any impact when setting up layers and tagsets via Inception?

No, not really. But I guess I still don't understand your problem.

So I understand you want to copy the structure from the quran corpus. I further understand that there are POS tags. I think you want that ADJ, PN, N and P are on a "word POS layer" but you want the PP to be on a "phrase POS layer", right? The dependencies on the word layer seem to work. In your second screenshot, I can also see a PP annotation in a different color (I assume that blue is the "word POS layer" and orange is the "phrase POS layer").

Let's stop here for a moment. If you set your "word POS layer" to allow stacking, then you shouldn't need the separate "phrase POS layer". But there may be reasons to have separate layers.

So let's see. Is the question why the orange PP is appearing below instead of above the N and P annotations? 

Having a quick glance at the ordering logic, I can see a comment there:

```
// - spans with more fragments go first
// - longer arc distances go last
// - spans with more arcs go last
// - compare the span widths,  put wider on bottom so they don't mess with arcs, or shorter on bottom if there are no arcs.
 // - if no other criterion is found, sort by type to maintain consistency
```

So it looks like it was a decision by the original brat authors to put wider spans closer to the text so there is more space for the arcs?

**Comment by Kentoseth** [comment]
> Otherwise, I am not sure which what you mean by "there again seems to be a bug on the display of Dependencies/PoS".

Sorry about this. This is a separate issue. I have split it out and it is now in #2686 

>  I think you want that ADJ, PN, N and P are on a "word POS layer"

The ADJ, PN, N and P are on the character-level layer. This is because of the morphology of Arabic. So I set it up on the character-level.

> but you want the PP to be on a "phrase POS layer", right?

Correct. But in Arabic you can have 2 scenarios with Preposition Phrases(PP). They are:

1. بِسۡمِ - a single word, but the preposition is attached to the noun("bi")
2. في بيت - a phrase, with the preposition on its own and noun on its own

> In your second screenshot, I can also see a PP annotation in a different color (I assume that blue is the "word POS layer" and orange is the "phrase POS layer").

To clarify firstly, the Quran Corpus seems to have the "dependency layer" in between the 2 PoS layers. 

This is what I was struggling to reproduce. I cannot see how to 'add' a PoS(the PP) layer 'above' or 'stacked' onto the Dependency Layer.

So what I did was to create the Arabic PoS Phrase Layer(where PP resides) just as I created the Arabic PoS word/character Layer without any stacking.

This is why you see them in different colours.

However, the Arabic PoS Phrase Layer(PP) is not "stacked above" the Arabic Dependency Layer in Inception but it is in the Quran Corpus. See the screenshot with red markings below.

![pic3_notes](https://user-images.githubusercontent.com/2989402/140641424-5c593d6b-67da-40e1-9885-6b8bc783b21a.png)

> Let's stop here for a moment. If you set your "word POS layer" to allow stacking, then you shouldn't need the separate "phrase POS layer". But there may be reasons to have separate layers.

Once again I apologize for the confusion. The issue might be that due to the way Arabic grammar/morphology is structured, it is a bit confusing to people not familiar with it(and many who are).

The long write-up about PP is here: https://corpus.quran.com/documentation/prepositionphrase.jsp

Link about Phrase Nodes: https://corpus.quran.com/documentation/dependencygraph.jsp#phrases

The tl;dr is:

1. في = preposition
2. بيت = noun
3. The relation of 2. to 1. is majrur
4. Together, they create the Preposition Phrase
5. This PP is then used as a relation to something else(muta'alliq)

You can see another example of this principle w.r.t Nominal Sentences(NS) here: https://corpus.quran.com/treebank.jsp?chapter=112&verse=1

If I am still not able to convey myself properly, the summary is:

I want to attach a PoS Phrase Layer on top of the Dependency Layer. The picture below summarizes it also:

![pic4_notes](https://user-images.githubusercontent.com/2989402/140643116-cbfadd63-ef39-4264-8d56-e677c4782a5c.png)
